File: app/src/main/java/com/quran/labs/androidquran/ui/preference/SeekBarPreference.java
/* The following code was written by Matthew Wiggins 
 * and is released under the APACHE 2.0 license 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package com.quran.labs.androidquran.ui.preference;

import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;
import android.widget.SeekBar;
import android.widget.TextView;

import com.quran.labs.androidquran.R;
import com.quran.labs.androidquran.data.Constants;
import com.quran.labs.androidquran.util.QuranUtils;

import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;

public class SeekBarPreference extends Preference implements SeekBar.OnSeekBarChangeListener {

  private static final String ANDROID_NS = "http://schemas.android.com/apk/res/android";

  private Context mContext;
  private SeekBar mSeekBar;
  private TextView mValueText;
  protected TextView mPreviewText;

  private String mSuffix;
  private int mTintColor;
  private int mCurrentValue;
  private int mDefault, mMax, mValue = 0;

  public SeekBarPreference(Context context, AttributeSet attrs) {
    super(context, attrs);
    mContext = context;
    mSuffix = attrs.getAttributeValue(ANDROID_NS, "text");
    mDefault = attrs.getAttributeIntValue(ANDROID_NS, "defaultValue",
        Constants.DEFAULT_TEXT_SIZE);
    mMax = attrs.getAttributeIntValue(ANDROID_NS, "max", 100);
    setLayoutResource(R.layout.seekbar_pref);
    mTintColor = ContextCompat.getColor(context, R.color.accent_color);
  }

  @Override
  public void onBindViewHolder(PreferenceViewHolder holder) {
    super.onBindViewHolder(holder);
    mSeekBar = (SeekBar) holder.findViewById(R.id.seekbar);
    mValueText = (TextView) holder.findViewById(R.id.value);
    mPreviewText = (TextView) holder.findViewById(R.id.pref_preview);
    mPreviewText.setVisibility(getPreviewVisibility());
    mSeekBar.setOnSeekBarChangeListener(this);
    styleSeekBar();
    mValue = shouldPersist() ? getPersistedInt(mDefault) : 0;
    mSeekBar.setMax(mMax);
    mSeekBar.setProgress(mValue);
  }

  /**
   * Visibility of the preview view under the seek bar
   */
  protected int getPreviewVisibility() {
    return View.GONE;
  }

  private void styleSeekBar() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
      final Drawable progressDrawable = mSeekBar.getProgressDrawable();
      if (progressDrawable != null) {
        if (progressDrawable instanceof LayerDrawable) {
          LayerDrawable ld = (LayerDrawable) progressDrawable;
          int layers = ld.getNumberOfLayers();
          for (int i = 0; i < layers; i++) {
            ld.getDrawable(i).mutate().setColorFilter(mTintColor, PorterDuff.Mode.SRC_ATOP);
          }
        } else {
          progressDrawable.mutate().setColorFilter(mTintColor, PorterDuff.Mode.SRC_ATOP);
        }
      }

      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
        final Drawable thumb = mSeekBar.getThumb();
        if (thumb != null) {
          thumb.mutate().setColorFilter(mTintColor, PorterDuff.Mode.SRC_ATOP);
        }
      }
    }
  }

  @Override
  protected void onSetInitialValue(@Nullable Object defaultValue) {
    super.onSetInitialValue(defaultValue);
    mValue = shouldPersist()
        ? getPersistedInt(mDefault)
        : (defaultValue != null ? (Integer) defaultValue : 0);
  }

  @Override
  public void onProgressChanged(SeekBar seek, int value, boolean fromTouch) {
    String t = QuranUtils.getLocalizedNumber(mContext, value);
    mValueText.setText(mSuffix == null ? t : t.concat(mSuffix));
    mCurrentValue = value;
  }

  @Override
  public void onStartTrackingTouch(SeekBar seek) {
  }

  @Override
  public void onStopTrackingTouch(SeekBar seek) {
    if (shouldPersist()) {
      persistInt(mCurrentValue);
      callChangeListener(mCurrentValue);
    }
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/preference/SeekBarPreference.kt
package com.quran.labs.androidquran.ui.preference

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.SeekBar
import android.widget.TextView
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.quran.labs.androidquran.R
import com.quran.labs.androidquran.data.Constants
import com.quran.labs.androidquran.util.QuranUtils

open class SeekBarPreference(
  context: Context,
  attrs: AttributeSet
) : Preference(context, attrs), SeekBar.OnSeekBarChangeListener {

  private lateinit var valueText: TextView
  protected lateinit var previewText: TextView
  protected lateinit var previewBox: View

  private val suffix = attrs.getAttributeValue(ANDROID_NS, "text")
  private val default = attrs.getAttributeIntValue(
    ANDROID_NS,
    "defaultValue",
    Constants.DEFAULT_TEXT_SIZE
  )
  private val maxValue = attrs.getAttributeIntValue(ANDROID_NS, "max", 100)
  private var currentValue = 0
  protected var value = 0

  init {
    layoutResource = R.layout.seekbar_pref
  }

  override fun onBindViewHolder(holder: PreferenceViewHolder) {
    super.onBindViewHolder(holder)
    val seekBar = holder.findViewById(R.id.seekbar) as SeekBar
    valueText = holder.findViewById(R.id.value) as TextView
    previewText = holder.findViewById(R.id.pref_preview) as TextView
    previewBox = holder.findViewById(R.id.preview_square)
    previewText.visibility = getPreviewVisibility()
    seekBar.setOnSeekBarChangeListener(this)
    value = if (shouldDisableView) getPersistedInt(default) else 0
    seekBar.apply {
      max = maxValue
      progress = value
    }
  }

  override fun onSetInitialValue(defaultValue: Any?) {
    super.onSetInitialValue(defaultValue)
    value = if (shouldPersist()) {
      getPersistedInt(default)
    } else {
      if (defaultValue != null) default else 0
    }
  }

  override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
    val t = QuranUtils.getLocalizedNumber(context, progress)
    valueText.text = if (suffix == null) t else "$t$suffix"
    currentValue = progress
  }

  override fun onStartTrackingTouch(seekBar: SeekBar?) {}

  override fun onStopTrackingTouch(seekBar: SeekBar?) {
    if (shouldPersist()) {
      persistInt(currentValue)
      callChangeListener(currentValue)
    }
  }

  /**
   * Visibility of the preview view under the seek bar
   */
  protected open fun getPreviewVisibility(): Int = View.GONE

  companion object {
    private const val ANDROID_NS = "http://schemas.android.com/apk/res/android"
  }
}
